### DISCUSSION TITLE
Break out PP calculations to a separate nuget module

### DISCUSSION THREAD

**Discussion by MaxKruse** [ppy/osu#16815] [Ideas] [answered]
Hi,

based on the variety of community tools that utilize their own re-implementations of the pp system, i think overall development burden and dependency of re-implementations every time can be lessened if the pp-calculations for each mod can be essentially managed in a central place.

Just a general idea, I am sure i am missing some considerable technical arguments against this.

**Comment by peppy** [comment, marked as answer]
I'm not sure I understand. Can you better describe the "development burden"? Because splitting out more dependencies sounds like *more* burden to me, not less.

**Reply by MaxKruse** [reply]
I meant the development burden of anybody that is not consuming, but instead re-implementing the current pp/difficulty calculations.

I was thinking that just how the framework is something to build osu on, so can difficulty/pp be something to build other tools on. I am aware this is by no means a priority, if at all accepted.

**Reply by peppy** [reply]
I still don't really follow. Can you describe precisely what you're trying to do?

For people working on pp, we already have a very streamlined setup for doing so. You can use [osu-tools](https://github.com/ppy/osu-tools) with a local osu checkout. Making nuget packages would add one extra local checkout requirement with little to no gains.

The only benefit I could think of off-hand would be having pull requests in a separate place, but hat can be done with labels.

**Reply by MaxKruse** [reply]
Ok, let me explain some use cases i would have, with how it would be needed to be implemented right now, and a possible solution:

1. A web-api that returns a pp value given mods, combo/hitcounts and a beatmap file, hash or id.

To achieve this, we currentlywould have to either use a local osu-tools binary, or fork osu-tools and modify that heavily. Consolidating and making a separate API for performance calculations, results in a variety of new services and toolings, for example retro-pp using older versions of a NuGet-Package (if so available). Implementing this single source-of-truth also offloads the actual computing to a separate program, so consumers like stream-companions, discord bots, etc. are not restricted by performance or tooling implementations of third parties.

2. Community tools for osu, for example gosumemory or McOsu, have to maintain their own fork of a re-implementation (often oppai), that itself is abandoned.

Community tools are a great part about osu, and looking at other games that utilize "3rd-party tooling" to improve the experience of players, i think a valid argument to make is to support this making of tooling or utilities by providing a simple programming interface to those that want to consume an official, maintained source of truth.

3. Have performance in one spot, and one spot alone. This makes the overall project structure of osu easier to grasp for individuals. As far as i can tell, there are currently 3? official projects that all re-implement the same basic algorithm: osu, osu-tools (which consumes components of osu and its rulesets), and osu-performance, which is an entire re-implementation in C++.

Having a single place that holds the implementation makes for a sort-of hub of development on it from outside sources. It would also reduce the needed development-overhead from bringing all variants of the algorithm to the same state. A single repository that serves for various includes/interfaces for other languages might be a small, but non-zero reduction to overhead in this aspect of the overall project.

I want to post this under the aspect that this suggestion helps not today, but in the future, to both improve community-tooling and slightly reduce development overhead. Increasing the version number in a project file is considerably less work than re-implementing the same changes to different projects.

**Reply by bdach** [reply]
> we currentlywould have to either use a local osu-tools binary, or fork osu-tools and modify that heavily

how else would you have that happen? even if the performance calculation was a separate nuget, then that is only an option for c# developers (and you can already consume ruleset nugets, which are the ones that contain the pp calc logic, i might add). people write their tools in python, javascript, whatever. at which point the nuget isn't sufficient anymore, you need FFI or something that wraps the nuget, such as... osu-tools.

> osu-tools (which consumes components of osu and its rulesets)

osu-tools does not "re-implement" anything. it is a wrapper over the osu! packages which allows to perform calculations more conveniently. there is no actual calculation logic in osu-tools itself. just glue code.

> osu-performance, which is an entire re-implementation in C++

that only still exists for legacy reasons and is being actively phased out.

**Reply by smoogipoo** [reply]
Let me take a different approach to this conversation, because I noticed your specific use of the "single repository that serves for various includes/interfaces for other languages" blurb.

Is that the part that you are actually interested in? If so, I think those could go directly into `osu-tools` - for example there could be an implementation for a web service to calc sr/pp (something similar to [osu-beatmap-difficulty-lookup-cache](https://github.com/ppy/osu-beatmap-difficulty-lookup-cache), or a local TCP server plus an API for local NodeJS/C++/etc apps, and so on.

Otherwise, I still don't understand what you want from the rest of this post.
